### PR TITLE
core auto recall

### LINKED ISSUE #206: Auto Recall
Implements:
- [x] Add recall percent in to the smart contract
- [ ] Recall jobs in core
- [ ] Recall history for fullnode tomox

### LINKED ISSUE #207: Get token Price in tomox lending
Implements:
- [ ] Insert price (with epoch number) in contract
- [ ] Update get price function: 
  - If tomox price = 0, get price in contract
  - If tomox price # 0, get price in contract (with comparing epoch number), if NOK, using TomoX price

### PR DESCRIPTION
**Pull request by nguyenbatam**
**Changes in this PR:** (resolve #206  resolve #207 )
- core auto recall  & get token price from contract
  - recall when liquidate Price=collateral Price *  100/ recall rate  ~ 50 %
  - update new liquidate Price =  collateral Price * liquidation rate / deposit rate ~ 73%
- Export recall history to database

Testing:
- Send 1 INVEST item, 1 BORROW item in the same pair (BTC/USDT) to generate a lending trade
- Run this script (within from block **checkpoint** to **checkpoint + 100**): tomochain/contracts/tomox/simulation/price/main.go
  - Update USDT/TOMO price to contract
  - Update BTC/TOMO price to contract

### CODE REVIEW

File: tomoxlending/order_processor.go [hunk 1]
```diff
@@ -957,3 +970,30 @@ func (l *Lending) ProcessTopUpLendingTrade(lendingStateDB *lendingstate.LendingS
 	log.Debug("ProcessTopUp successfully", "price", newLiquidationPrice, "lockAmount", newLockedAmount)
 	return nil, false, &newLendingTrade
 }
+
+func (l *Lending) ProcessRecallLendingTrade(lendingStateDB *lendingstate.LendingStateDB, statedb *state.StateDB, tradingStateDb *tradingstate.TradingStateDB, lendingTradeId common.Hash, lendingBook common.Hash, newLiquidationPrice *big.Int) (error, bool, *lendingstate.LendingTrade) {
+	log.Debug("ProcessRecallLendingTrade", "lendingTradeId", lendingTradeId.Hex(), "lendingBook", lendingBook.Hex(), "newLiquidationPrice", newLiquidationPrice)
+	lendingTrade := lendingStateDB.GetLendingTrade(lendingBook, lendingTradeId)
+	if lendingTrade == lendingstate.EmptyLendingTrade {
+		return fmt.Errorf("process deposit for emptyLendingTrade is not allowed. lendingTradeId: %v", lendingTradeId.Hex()), true, nil
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: update error message:
return fmt.Errorf("**processRecall** for emptyLendingTrade is not allowed. lendingTradeId: %v", lendingTradeId.Hex()), true, nil

File: tomoxlending/tomoxlending.go [hunk 2]
```diff
@@ -792,7 +792,29 @@ func (l *Lending) ProcessLiquidationData(chain consensus.ChainContext, time *big
...
+		for lowestLiquidatePrice.Sign() > 0 && recalLiquidatePrice.Cmp(lowestLiquidatePrice) > 0 {
+			for lendingBook, tradingIds := range liquidationData {
+				for _, tradingIdHash := range tradingIds {
+					trade := lendingState.GetLendingTrade(lendingBook, tradingIdHash)
+					if trade.AutoTopUp {
+						if err, _, newTrade := l.ProcessRecallLendingTrade(lendingState, statedb, tradingState, lendingBook, tradingIdHash, recalLiquidatePrice); err == nil {
+							// if this action complete successfully, do not liquidate this trade in this epoch
+							log.Debug("AutoRecall", "borrower", trade.Borrower.Hex(), "collateral", newTrade.CollateralToken.Hex(), "newLockedAmount", newTrade.CollateralLockedAmount)
+							finalizedTrades[newTrade.Hash] = newTrade
+							continue
```
Comment (thanhnguyennguyen) [inline tomoxlending/tomoxlending.go]: no need **continue** here

File: core/lending_pool_test.go [hunk 3]
```diff
@@ -168,8 +168,8 @@ func testSendLending(key string, nonce uint64, lendToken, collateralToken common
 }
 
 func TestSendLending(t *testing.T) {
-	t.SkipNow() //TODO: remove it to run this test
-	key := ""
+	//t.SkipNow() //TODO: remove it to run this test
+	key := "65ec4d4dfbcac594a14c36baa462d6f73cd86134840f6cf7b80a1e1cd33473e2"
```
Comment (thanhnguyennguyen) [inline core/lending_pool_test.go]: don't commit them

File: tomoxlending/order_processor.go [hunk 4]
```diff
@@ -61,6 +62,12 @@ func (l *Lending) ApplyOrder(createdBlockTime uint64, coinbase common.Address, c
 	}
 
 	switch order.Type {
+	case lendingstate.LendingStatusCancelled:
+		err, reject := l.ProcessCancelOrder(header, lendingStateDB, statedb, tradingStateDb, chain, coinbase, lendingOrderBook, order)
+		if err != nil || reject {
+			rejects = append(rejects, order)
+		}
+		return trades, rejects, nil
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: rebase mistake: remove them, please
Cancel items are processed here https://github.com/tomochain/tomochain/pull/211/files#diff-bb070402da3cf4811b1d48bdf02391e6R90

**Review by thanhnguyennguyen — CHANGES_REQUESTED** [review]

File: tomoxlending/tomoxlending.go [hunk 5]
```diff
@@ -799,7 +800,31 @@ func (l *Lending) ProcessLiquidationData(chain consensus.ChainContext, time *big
 					}
 				}
 			}
-			highestPrice, liquidationData = tradingState.GetHighestLiquidationPriceData(orderbook, liquidationPrice)
+			highestLiquidatePrice, liquidationData = tradingState.GetHighestLiquidationPriceData(orderbook, collateralPrice)
+		}
+		// recall trades
+		depositRate, liquidationRate, _, recallRate, _ := lendingstate.GetCollateralDetail(statedb, lendingPair.CollateralToken)
+		recalLiquidatePrice := new(big.Int).Mul(collateralPrice, common.BaseRecall)
+		recalLiquidatePrice = new(big.Int).Div(collateralPrice, recallRate)
```
Comment (thanhnguyennguyen) [inline tomoxlending/tomoxlending.go]: recalLiquidatePrice = new(big.Int).Div(**recalLiquidatePrice**, recallRate)

**Review by thanhnguyennguyen — CHANGES_REQUESTED** [review]

**Review by thanhnguyennguyen — APPROVED** [review]

**Review by thanhson1085 — APPROVED** [review]

**Pull request merged**